conf_file.h

```diff
@@ -17,24 +17,123 @@
+     * @storage_offset The offset from the beggining of the struct used in extract_options() to
+     * store the values.
+     * @storage_len The length in bytes of the storage poited by @a storage_offset.
+     */
+    struct OptionsTable {
+        const char *key;
+        bool required;
+        int (*parser_func)(const char *val, size_t val_len, void *storage, size_t storage_len);
+        off_t storage_offset;
+        size_t storage_len;
+    };
+
+    /**
+     * Load and parse a configuration file.
+     *
+     * Open the file using mmap and parse it, keeping an internal structure with pointers to conf
+     * sections and options, in order to do a quick extraction using extract_options().
+     * When an option in a section is already present before parse is called, the option is
+     * overrided by the option from the new file. Other options from sections are not changed.
+     * Files are kept open until release_all() is called or the ConfFile object is destructed.
+     *
+     * @param filename The conf filename
+     * @return errno on IO or parsing errors or @c 0 if successful
+     */
+    int parse(const char *filename);
+
+    /**
+     * Release all opened files and internal structures from this ConfFile.
+     */
+    void release_all();
+
+    /**
+     * Extract options set in @a table from section @a section_name
+     *
+     * @param section_name Name of the section to extract options.
+     * @param table Array of TableOption with information on which fields are going to be
+     * extracted and how to extract them.
+     * @param table_len The number of elements in @a table.
+     * @param data A pointer to the struct that will be used to hold the extracted data.
+     */
+    int extract_options(const char *section_name, const OptionsTable table[], size_t table_len,
+                        void *data);
+    /**
+     * Loop to all sub-sections from the section @a section_name
+     *
+     * @param section_name The name of the section to look for sub-sections
+     * @param separator Charatecter that separates a section from the sub-section
+     * @param sub_name A pointer to a char* that will be updated with current sub-section name
+     * @param sub_len A pointer to size_t element be filled with the length of the current
+     * sub-section name.
+     * @param table Array of TableOption with information on which fields are going to be
+     * extracted and how to extract them.
+     * @param table_len The number of elements in @a table.
+     * @param data A pointer to the struct that will be used to hold the extracted data.
+     * @param ptr A void pointer to be used internally in the loop. Must be initialized with @c NULL
+     * for first execution.
+     */
+    int loop_sub_sections(const char *section_name, char separator, const char **sub_name,
+                          size_t *sub_len, const OptionsTable table[], size_t table_len, void *data,
+                          void **ptr);
```

Comment:
I'm not sure about the approach on this function. I'd rather try to reuse extract_options() above, which would make most of these arguments go away.  The idea would be to have a `get_sections(const char *glob)` that accepts a glob and returns an iterator. You then loop on this iterator and call `extract_options()`. The "sub section" name is saved on this iterator.

I'd also avoid the "sub section" nomenclature. It's just a section IMO.